### ISSUE TITLE
K8s-dns found cluster mode invalid

### ISSUE THREAD

**Issue by xiyangxdy** [emqx/emqx#2575]
#### Environment

- OS: linux
- Erlang/OTP: 
- EMQ: 2.3.11

#### Description
     

  I want to use k8s-dns to discover the cluster pattern, but it can't be formed with the following configuration. After entering the pod, the node name is named after the ip of the pod. And after looking at the /etc/hosts file, I think the node name should be node.name = emqtt@emqtt-6bsrt. So I tried to add EMQX_NODE_NAME in yaml, but the pod name is dynamically generated, I have not configured successfully. Have you had a similar experience?

**yaml**
```
apiVersion: v1
kind: ReplicationController
metadata:
  name: emqtt
  labels:
    name: emqtt
spec:
  replicas: 3
  selector:
    name: emqtt
  template:
    metadata:
      labels:
        name: emqtt
    spec:
      containers:
      - name: emqtt
        image: emqtt:v2.3
        ports:
        - containerPort: 80
        env:
          - name: EMQ_NODE__DIST_NET_TICKTIME
            value: "300"
          - name: EMQ_NAME
            value: "emqtt"
          - name: EMQ_CLUSTER__DISCOVERY
            value: "k8s"
          - name: EMQ_CLUSTER__K8S__APISERVER
            value: "http:\\/\\/***:8080"
          - name: EMQ_CLUSTER__K8S__SERVICE_NAME
            value: "emqtt"
          - name: EMQ_CLUSTER__K8S__ADDRESS_TYPE
            value: "dns"
          - name: EMQ_CLUSTER__K8S__APP_NAME
            value: "emqtt"
          - name: EMQ_CLUSTER__NAME
            value: "emqttcl"
          - name: EMQ_CLUSTER__AUTOHEAL
            value: "on"
          - name: EMQ_CLUSTER__AUTOCLEAN
            value: "5m"
          - name: EMQ_NODE__COOKIE
            value: "secretcookieeeetfggggg"
          #- name: EMQ_NODE_NAME
          #  value: emq@${POD_NAME}
        volumeMounts:
          - name: syslog
            mountPath: /workspace/logs
          - name: systime
            readOnly: true
            mountPath: /etc/localtime
      volumes:
        - name: syslog
          hostPath: {path: /home/azureuser/logs}
        - name: systime
          hostPath: {path: /etc/localtime}

```

**pod name——emqtt-6bsrt** 

**pod /etc/hosts**
```
# Kubernetes-managed hosts file.
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
fe00::0 ip6-mcastprefix
fe00::1 ip6-allnodes
fe00::2 ip6-allrouters
172.34.26.102   emqtt-6bsrt
172.34.26.102        172.34.26.102
```

**emq.conf**
`## Node name
node.name = emqtt@172.34.26.102 `

*A description of the issue*

**Comment by zhanghongtong** [comment]
This is a real problem, we are working on it, thank you for your feedback.

**Comment by xiyangxdy** [comment]
> This is a real problem, we are working on it, thank you for your feedback.
       嘿嘿，还是说中文吧！
       后来我了解到EMQ如果要用dns模式部署的话，node.name的命名应该为emqx@10-244-1-3.default.pod.cluster.local 这种格式
这个10-244-1-3 其实就是pod的ip转化过来的吧。有个问题就是，如果删除某一个EMQ节点，这个pod ip是会变的，所以这个域名也会变，其他节点存储的节点信息会变化，我觉得这样的话很有可能集群脑裂。事实证明，我用ip模式部署EMQ集群，脑裂过大概3次，问过你们的官方人员，每次都没有一个结论。或许这是一个BUG？
        我也比较了解rabbitmq，他们用dns模式部署的话，同时用的statefulset资源类型（EMQ则用的是ReplicationController或者deployment），这个node-name是永远不变的，比如rabbitmq的一个pod名字为rabbitmq-0，域名为rabbit@rabbitmq-0.rabbitmq.default.svc.cluster.local。所以我希望你们可以学习下rabbitmq这种部署方式。
       最重要的是，我在k8s上部署过rabbitmq ip和dns模式，dns模式比ip模式稳定得许多，EMQ也是Erlang写的，或许dns模式更适合它呢？

**Comment by xiyangxdy** [comment]
> > This is a real problem, we are working on it, thank you for your feedback.
> 
> ```
>    嘿嘿，还是说中文吧！
>    后来我了解到EMQ如果要用dns模式部署的话，node.name的命名应该为emqx@10-244-1-3.default.pod.cluster.local 这种格式
> ```
> 
> 这个10-244-1-3 其实就是pod的ip转化过来的吧。有个问题就是，如果删除某一个EMQ节点，这个pod ip是会变的，所以这个域名也会变，其他节点存储的节点信息会变化，我觉得这样的话很有可能集群脑裂。事实证明，我用ip模式部署EMQ集群，脑裂过大概3次，问过你们的官方人员，每次都没有一个结论。或许这是一个BUG？
> 我也比较了解rabbitmq，他们用dns模式部署的话，同时用的statefulset资源类型（EMQ则用的是ReplicationController或者deployment），这个node-name是永远不变的，比如rabbitmq的一个pod名字为rabbitmq-0，域名为[rabbit@rabbitmq-0.rabbitmq.default.svc.cluster.local](mailto:rabbit@rabbitmq-0.rabbitmq.default.svc.cluster.local)。所以我希望你们可以学习下rabbitmq这种部署方式。
> 最重要的是，我在k8s上部署过rabbitmq ip和dns模式，dns模式比ip模式稳定得许多，EMQ也是Erlang写的，或许dns模式更适合它呢？

**Comment by zhanghongtong** [comment]
@xiyangxdy thank you for your reply
+ Regarding the problem of k8s-dns, it is really necessary to change the node-name of emqx to the format "emqx@10-244-1-3.default.pod.cluster.local". In addition, there is a problem with our code. Caused the cluster to fail, we will fix this in the latest release.
+ Regarding rabbitmq, we will consider your suggestion later, thank you very much.
+ Regarding the brain splitting problem of using IP clusters, we are more concerned about it. Can you provide more detailed information? For example, is your k8s network plugin `calico` or `flannel`?

**Comment by xiyangxdy** [comment]
> @xiyangxdy thank you for your reply
> 
> * Regarding the problem of k8s-dns, it is really necessary to change the node-name of emqx to the format "[emqx@10-244-1-3.default.pod.cluster.local](mailto:emqx@10-244-1-3.default.pod.cluster.local)". In addition, there is a problem with our code. Caused the cluster to fail, we will fix this in the latest release.
> * Regarding rabbitmq, we will consider your suggestion later, thank you very much.
> * Regarding the brain splitting problem of using IP clusters, we are more concerned about it. Can you provide more detailed information? For example, is your k8s network plugin `calico` or `flannel`?

calico

**Comment by zhanghongtong** [comment]
@xiyangxdy thank you for your reply
We have been experimenting with `flannel` all the time, we will try to reproduce it in `calico`

**Comment by zhanghongtong** [comment]
@xiyangxdy I have used calico as a network plugin and still haven't reproduce the problem of cluster splitting. Is there anything else I should know?

**Comment by xiyangxdy** [comment]
@zhanghongtong This problem is not necessarily the case. The environment on my side is after a period of operation, suddenly cracked, no omen and human operation.This problem is also very tricky for us, we have to use a single EMQ in production to maintain stability

**Comment by zhanghongtong** [comment]
@xiyangxdy I just consulted our developers. Our cluster split recovery feature requires that the number of nodes in the cluster must be singular to work properly, and if the number of nodes is double, there is no recovery. Is your number of nodes singular?

**Comment by xiyangxdy** [comment]
@zhanghongtong Yes, the number of our cluster nodes is singular. I have looked for your developers( Gilbert ) to identify the problem, gave him some logs, maybe you can ask him to see

**Comment by HJianBo** [comment]
Hi, guys. This issue has fixed at https://github.com/emqx/ekka/pull/53 and released at v3.1.2. Please check it!

I'll close it first. If any problems there, reopen it!
